From 36060204568f756ac459914356ae060739958960 Mon Sep 17 00:00:00 2001
From: Johnny Mnemonic <jm@machine-hall.org>
Date: Sat, 14 Feb 2026 16:13:45 +0100
Subject: [PATCH 100/101] ia64: drop zero_page_memmap_ptr and paging_init()

d49004c5f0c140bb83c87fab46dcf449cf00eb24 and/or the then required ia64
adaptation for it created an issue for HP-Sim/Ski kernels running in Ski
that prevented successful execution of userland programs:

```
[...]
[    1.194139] scsi host0: simulated SCSI host adapter
[    1.194139] scsi 0:0:0:0: Direct-Access     HP       SIMULATED DISK   0.00 PQ: 0 ANSI: 2
[    1.225399] sd 0:0:0:0: [sda] 16777216 512-byte logical blocks: (8.59 GB/8.00 GiB)
[    1.225399] sd 0:0:0:0: [sda] Write Protect is off
[    1.225399] sd 0:0:0:0: [sda] Asking for cache data failed
[    1.225399] sd 0:0:0:0: [sda] Assuming drive cache: write through
[    1.256619] sd 0:0:0:0: [sda] Attached SCSI disk
[    1.412845] check access for rdinit=/init failed: -2, ignoring
[    2.037829] EXT4-fs (sda): mounted filesystem 590ca40f-256f-4a23-927b-fff68d58bcf5 r/w with ordered data mode. Quota mode: disabled.
[    2.037829] VFS: Mounted root (ext4 filesystem) on device 8:0.
[    2.037829] VFS: Pivoted into new rootfs
[    2.037829] Freeing unused kernel memory: 768K
[    2.037829] This architecture does not have kernel memory protection.
[    2.037829] Run /init as init process
[    2.069161] Kernel panic - not syncing: Attempted to kill init! exitcode=0x00000004
[    2.069161] ---[ end Kernel panic - not syncing: Attempted to kill init! exitcode=0x00000004 ]---
```

It seems like the prior use of the struct page pointer
zero_page_memmap_ptr had a play into this issue. Turns out, we can do w/o
it and instead define ZERO_PAGE(vaddr) with the conversion used in
paging_init() in mm/[dis]contig.c directly. Also drop paging_init()
because it is empty now.

This fixes the issue mentioned above **and** keeps things working on
real machines (tested on rx2620, rx4640, rx2660, rx6600 and rx2800 i2).

General idea from 18b7cc70dea8b4071ebf736724e99103115c5f95.
---
 arch/ia64/include/asm/pgtable.h |  4 +---
 arch/ia64/kernel/setup.c        |  1 -
 arch/ia64/mm/contig.c           | 10 ----------
 arch/ia64/mm/discontig.c        |  8 --------
 arch/ia64/mm/init.c             |  3 ---
 5 files changed, 1 insertion(+), 25 deletions(-)

diff --git a/arch/ia64/include/asm/pgtable.h b/arch/ia64/include/asm/pgtable.h
index 7542a4badf06..a6a7ccab977e 100644
--- a/arch/ia64/include/asm/pgtable.h
+++ b/arch/ia64/include/asm/pgtable.h
@@ -397,7 +397,6 @@ pte_same (pte_t a, pte_t b)
 #define update_mmu_cache(vma, address, ptep) do { } while (0)
 
 extern pgd_t swapper_pg_dir[PTRS_PER_PGD];
-extern void paging_init (void);
 
 /*
  * Encode/decode swap entries and swap PTEs. Swap PTEs are all PTEs that
@@ -444,8 +443,7 @@ static inline pte_t pte_swp_clear_exclusive(pte_t pte)
  * for zero-mapped memory areas etc..
  */
 extern unsigned long empty_zero_page[PAGE_SIZE/sizeof(unsigned long)];
-extern struct page *zero_page_memmap_ptr;
-#define ZERO_PAGE(vaddr) (zero_page_memmap_ptr)
+#define ZERO_PAGE(vaddr)	(virt_to_page(ia64_imva(empty_zero_page)))
 
 /* We provide our own get_unmapped_area to cope with VA holes for userland */
 #define HAVE_ARCH_UNMAPPED_AREA
diff --git a/arch/ia64/kernel/setup.c b/arch/ia64/kernel/setup.c
index 3dbd37da666a..21584484a05a 100644
--- a/arch/ia64/kernel/setup.c
+++ b/arch/ia64/kernel/setup.c
@@ -642,7 +642,6 @@ setup_arch (char **cmdline_p)
 #endif
 
 	screen_info_setup();
-	paging_init();
 
 	clear_sched_clock_stable();
 }
diff --git a/arch/ia64/mm/contig.c b/arch/ia64/mm/contig.c
index 9a2f33fc843e..fb3c556c3d6b 100644
--- a/arch/ia64/mm/contig.c
+++ b/arch/ia64/mm/contig.c
@@ -196,13 +196,3 @@ void __init arch_zone_limits_init(unsigned long *max_zone_pfns)
 
 	verify_gap_absence();
 }
-
-/*
- * Initialize the kernel's zero page.
- */
-
-void __init
-paging_init (void)
-{
-	zero_page_memmap_ptr = virt_to_page(ia64_imva(empty_zero_page));
-}
diff --git a/arch/ia64/mm/discontig.c b/arch/ia64/mm/discontig.c
index 8592bf896c68..099ebaef58e2 100644
--- a/arch/ia64/mm/discontig.c
+++ b/arch/ia64/mm/discontig.c
@@ -594,14 +594,6 @@ void __init arch_zone_limits_init(unsigned long *max_zone_pfns)
 	max_zone_pfns[ZONE_NORMAL] = max_low_pfn;
 }
 
-/*
- * Initialize the kernel's zero page.
- */
-void __init paging_init(void)
-{
-	zero_page_memmap_ptr = virt_to_page(ia64_imva(empty_zero_page));
-}
-
 /* workaround for a warning with -Wmissing-prototypes */
 pg_data_t * __init arch_alloc_nodedata(int nid);
 void arch_refresh_nodedata(int update_node, pg_data_t *update_pgdat);
diff --git a/arch/ia64/mm/init.c b/arch/ia64/mm/init.c
index 401bbe14ecbb..f6e8f2de2ba4 100644
--- a/arch/ia64/mm/init.c
+++ b/arch/ia64/mm/init.c
@@ -43,9 +43,6 @@ extern void ia64_tlb_init (void);
 
 unsigned long MAX_DMA_ADDRESS = PAGE_OFFSET + 0x100000000UL;
 
-struct page *zero_page_memmap_ptr;	/* map entry for zero page */
-EXPORT_SYMBOL(zero_page_memmap_ptr);
-
 void
 __ia64_sync_icache_dcache (pte_t pte)
 {
-- 
2.25.1

